apps/designer/app/canvas/hovered-instance-connector.ts

```diff
@@ -0,0 +1,101 @@
+import { useEffect } from "react";
+import debounce from "lodash.debounce";
+import { idAttribute } from "@webstudio-is/react-sdk";
+import { utils } from "@webstudio-is/project";
+import { publish, subscribe } from "~/shared/pubsub";
+import { useRootInstance } from "~/shared/nano-states";
+import { subscribeScrollState } from "~/shared/dom-hooks";
+
+export const useHoveredInstanceConnector = () => {
+  const [rootInstance] = useRootInstance();
+
+  useEffect(() => {
+    if (rootInstance === undefined) {
+      return;
+    }
+
+    let hoveredElement: undefined | Element = undefined;
+
+    const publishHover = debounce((element: Element) => {
+      const id = element.getAttribute(idAttribute) ?? undefined;
+      if (id === undefined) {
+        return;
+      }
+      const instance = utils.tree.findInstanceById(rootInstance, id);
+      if (instance === undefined) {
+        return;
+      }
+      // @todo combine hover events
+      publish({
+        type: "hoveredInstanceRect",
+        payload: element.getBoundingClientRect(),
+      });
+      publish({
+        type: "hoverInstance",
+        payload: {
+          id: instance.id,
+          component: instance.component,
+        },
+      });
+    }, 50);
+
+    const handleMouseOver = (event: MouseEvent) => {
+      const element = event.target;
+      if (element instanceof Element) {
+        // store hovered element locally to update outline when scroll ends
+        hoveredElement = element;
+        publishHover(element);
+      }
+    };
+
+    const handleMouseOut = () => {
+      hoveredElement = undefined;
+      publish({
+        type: "hoverInstance",
+        payload: undefined,
+      });
+    };
+
+    const eventOptions = { passive: true };
+    window.addEventListener("mouseover", handleMouseOver, eventOptions);
+    window.addEventListener("mouseout", handleMouseOut, eventOptions);
+
+    const unsubscribeNavigatorHoveredInstance = subscribe(
+      "navigatorHoveredInstance",
+      (navigatorHoveredInstance) => {
+        if (navigatorHoveredInstance === undefined) {
+          return;
+        }
+        const id = navigatorHoveredInstance.id;
+        const element =
+          document.querySelector(`[${idAttribute}="${id}"]`) ?? undefined;
+        if (element !== undefined) {
+          publishHover(element);
+        }
+      }
+    );
+
+    // remove hover outline when scroll starts
+    // and show it with new rect when scroll ends
+    const unsubscribeScrollState = subscribeScrollState({
+      onScrollStart() {
```

Comment:
I have feeling that it executed many times because of bug in scroll subsription below